### PR TITLE
A single framework for gen-kill and generic dataflow problems

### PR DESCRIPTION
**Pull request by ecstatic-morse**
This is the prototype implementation discussed in rust-lang/compiler-team#202. You can read a high-level description of it in [the proposal](https://hackmd.io/@39Qr_z9cQhasi25sGjmFnA/Skvd9rztS) for that design meeting. This would eventually supersede the existing `BitDenotation` interface.

r? @ghost 

cc @rust-lang/compiler (esp. @eddyb and @pnkfelix)

### CODE REVIEW

**Comment by JohnCSimon** [comment]
Ping from triage:
@ecstatic-morse Hi! is this PR still in progress?
thanks.

**Comment by JohnCSimon** [comment]
Pinging from triage:
@ecstatic-morse Can you please provide status on this PR?
thanks.

**Comment by ecstatic-morse** [comment]
This was discussed in the design meeting on Nov. 8 (rust-lang/compiler-team#202). The consensus was to move forward on a design similar to the prototype. I'll be actively working to merge this, hopefully with some input from the compiler team.

**Comment by ecstatic-morse** [comment]
Okay, this converts a commonly used dataflow analysis, `MaybeInitializedLocals` to the new framework to get an idea of usability/perf changes. The `Engine` implementation is now much simpler than the original since the choice of whether to cache block transfer functions is dynamic. This means we will be able to skip this for acyclic MIR in the future.

I've triggered a perf run and the results look okay but there is a slight regression in `wg-grammar`. I will try lifting the check for a cached block transfer function out of the loop to see if that helps. However, I think this is ready for an initial round of review.

r? @pnkfelix cc @rust-lang/compiler (feel free to reassign)

Here are the unresolved "details" questions from the design meeting that are pertinent to this PR along with my thoughts on each. We should resolve these questions before merging.
 - How do we feel about the `{before_,}*_effect` naming convention? It takes a bit of getting used to, and doesn’t extend nicely to backwards dataflow.
    - OTOH, having a canonical `statement_effect` which most implementers want to use and a secondary, prefixed effect which is only rarely used is good.
- Should we be passing a `mir::Statement` to the `statement` effect methods? What about `mir::Body` to `bits_per_block`?
    - I'd like to have a `HasBody<'mir, 'tcx>` trait that is a requirement for `AnalysisDomain` instead of passing in a separate body explicitly, but this will increase the amount of boilerplate required.
- Should the `Analysis` methods take `&self` or `&mut self`?
    - Some may depend on the results of an earlier dataflow analysis and want to hold a cursor.
    - However, transfer functions should be pure for a given MIR, so maybe forcing those analyses to use interior mutability is appropriate.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@ecstatic-morse - can you please fix the merge conflicts?

**Comment by ecstatic-morse** [comment]
@pnkfelix This is waiting for review. Feel free to reassign it. I could split out the unit test for `ResultsCursor` and the conversion of `MaybeInitializedPlaces` to the new framework if the diff is too imposing (I included it here so I could measure perf).

**Comment by ecstatic-morse** [comment]
@bors try
@rust-timer queue

**Comment by ecstatic-morse** [comment]
Seems there's no perf info for this this commit.

@rust-timer queue

**Comment by lqd** [comment]
@ecstatic-morse they are in the queue — https://perf.rust-lang.org/status.html — the server had an issue earlier and some data needed to be recomputed IIUC.

**Comment by ecstatic-morse** [comment]
Whoops, I see it now.

@bors try-

File: src/librustc_mir/dataflow/generic/mod.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        args: &[mir::Operand<'tcx>],
+        return_place: &mir::Place<'tcx>,
+    ) {
+        self.call_return_effect(state, block, func, args, return_place);
+    }
+}
+
+/// The legal operations for a transfer function in a gen/kill problem.
+pub trait GenKill<T>: Sized {
+    /// Inserts `elem` into the `gen` set, removing it the `kill` set if present.
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/mod.rs]: typo: removing it *from* the `kill` set if present

File: src/librustc_mir/dataflow/generic/mod.rs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    fn gen(&mut self, elem: T);
+
+    /// Inserts `elem` into the `kill` set, removing it the `gen` set if present.
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/mod.rs]: same typo: removing it *from* the `gen` set if present

File: src/librustc_mir/dataflow/generic/mod.rs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        func: &mir::Operand<'tcx>,
+        args: &[mir::Operand<'tcx>],
+        return_place: &mir::Place<'tcx>,
+    ) {
+        self.call_return_effect(state, block, func, args, return_place);
+    }
+}
+
+/// The legal operations for a transfer function in a gen/kill problem.
+pub trait GenKill<T>: Sized {
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/mod.rs]: I'm only on the first commit, so I don't really know the answer to this question yet: Does abstracting `GenKill<T>` into a trait make sense? Its hard to imagine a significantly different alternative implementation to what the existing `GenKillSet` struct does, apart from the details of how the two bitsets are represented, I suppose.

Comment (pnkfelix) [reply src/librustc_mir/dataflow/generic/mod.rs]: Okay I reached the end of this commit and now see that you have implemented `GenKill` for `BitSet`. I do not think this is a good idea.

Specifically, you have not stated in the documentation for `trait GenKill` that the underlying gen or kill sets may be *missing*, depending on how the trait is implemented. Having `fn kill()` just remove the bit from the gen-set does not really capture the semantics of a dataflow analysis: the clients parameterized over `GK: GenKill` are going to have to say "You must use `GenKillSet` for this to work", or they are going to have to ensure that their particular analysis does not actually rely ever generating a kill effect (in which case, I would have `<BitSet as GenKill>::kill` panic instead of solely clearing the bit in the gen-set.

File: src/librustc_mir/dataflow/generic/mod.rs (hunk 1 above)
Comment (pnkfelix) [reply src/librustc_mir/dataflow/generic/mod.rs]: also, the phrase "if present" is slightly ambiguous. 

When I first read "if present", I thought the phrase ("clearly") was referring to the `elem`, or rather the bit representing `elem` in the `kill` set. Which made sense to me.

But now that I have read the whole commit, I realize that it is possible that the "if present" was actually referring o the whole `kill`-set: i.e., this was an implicit note that the `kill`-set might be missing. (And likewise, the "if present" in `fn kill` below might be referring to the whole `gen`-set.)

I do not know what your actual intention was. I have already stated that I think it is a mistake to have `impl GenKill for BitSet`. If your intention matched my original interpretation, then maybe you can rephrase this as "Inserts `elem` into the `gen` set, removing it from the `kill` set if `elem` was present there." Or you could just say "Inserts `elem` into the `gen` set and removes it from the `kill` set", and rely on the reader to understand that under usual set theory, removing a non-present element is the identity function on sets.

File: src/librustc_mir/dataflow/move_paths/mod.rs [hunk 4]
```diff
@@ -72,6 +72,40 @@ impl<'tcx> MovePath<'tcx> {
 
         parents
     }
+
+    /// Looks for the first child of `self` for which `f` returns `true`.
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/move_paths/mod.rs]: I often read the term "child" as meaning "immediate child", as opposed to recursively traversing the family tree.

 1. Would "descendant" be better?
 2. Regardless of whether you change the terminology, you should state in the documentation the search strategy. I.e., If you say "Looks for the first descendant (under breadth-first search) for which ...", then I think it will be 100% clear that this is not restricting its search to the immediate children (because one would not distinguish DFS/BFS for a traversal of the immediate children).

File: src/librustc_mir/dataflow/generic/mod.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc_mir/dataflow/generic/mod.rs]: (I suppose another way it could hypothetically make sense to `impl GenKill for BitSet` would be for an analysis where the kill-set is *always* the set-complement of the gen-set. But AFAICT that is not what you implemented, and I am not aware of any analysis where the gen/kill sets satisfy that property.)

File: src/librustc_mir/borrow_check/mod.rs [hunk 5]
```diff
@@ -185,15 +185,11 @@ fn do_mir_borrowck<'a, 'tcx>(
...
-        &attributes,
-        &dead_unwinds,
-        MaybeInitializedPlaces::new(tcx, &body, &mdpe),
-        |bd, i| DebugFormatted::new(&bd.move_data().move_paths[i]),
-    ));
+    let flow_inits = MaybeInitializedPlaces::new(tcx, &body, &mdpe);
+    let mut flow_inits =
+        dataflow::generic::Engine::new_gen_kill(tcx, &body, def_id, &dead_unwinds, flow_inits)
+            .iterate_to_fixpoint()
+            .into_cursor(&body);
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/mod.rs]: OMG this is so much nicer to read than what I had before

File: src/librustc_mir/dataflow/generic/tests.rs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    fn block(&self) -> BasicBlock {
+        use SeekTarget::*;
+
+        match *self {
+            BlockStart(block) => block,
+            Before(loc) | After(loc) | AfterAssumeCallReturns(loc) => loc.block,
+        }
+    }
+
+    /// Returns the largest index (not including the basic block index) that should be set when
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/tests.rs]: I'm trying to understand the docs here, and this method + test-harness in general.

 * I have inferred that the indexes you are talking about are points before and after statements: { } S1 { }, { } S2 { }, ... etc. (At first I wasn't sure why you were distinguishing the point after S1 from the point before S2, but soon I realized that you are identifying points in the whole MIR, not just within a given basic block, and therefore in general you have to deal with control-flow splits and merges.)

 * When you say "that should be set", where are these indices being "*set*"? Is that referring to the `Dummy` that is storing the indices in a set?

 * Why is this called `effect` ? Because it is describing the "effect" of calling `seek` with a given `SeekTarget`? Or just that it describes the effect we want to compute on the bitsets for `Dummy`?

(I suppose part of my problem here is that I started reading this file from top to bottom, and it seems like the important explanations are further down, with `Dummy` and its related impls.)

File: src/librustc_mir/dataflow/generic/tests.rs [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            ret[bb].insert(BASIC_BLOCK_OFFSET + bb.index());
+        }
+
+        ret
+    }
+
+    /// Returns the expected state at the given `SeekTarget`.
+    ///
+    /// This is the union of index of the target basic block, the index assigned to the of the
+    /// target statement or terminator, and the indices of all preceding statements in the MIR.
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/tests.rs]: when you say "indices of all preceding statements in the MIR", does that mean:

 * all preceding statements *within* the target basic basic block (but not other blocks), or
 * all preceding statements both in the target basic block and elsewhere
    * if this, then is the "preceding" just talking about the order the statements are given in the MIR array,
    * or does "preceding" refer to control flow in some manner (but presumably some manner that ignores backward edges)?

File: src/librustc_mir/dataflow/generic/tests.rs [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+            is_cleanup: false,
+        })
+    };
+
+    let dummy_place = mir::Place {
+        base: mir::PlaceBase::Local(mir::RETURN_PLACE),
+        projection: ty::List::empty(),
+    };
+
+    block(4, mir::TerminatorKind::Return);
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/tests.rs]: I'm surprised there's no block with 0 statements (i.e. just a terminator) in included in this test.

File: src/librustc_mir/dataflow/generic/tests.rs [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        ret
+    }
+
+    /// Returns the expected state at the given `SeekTarget`.
+    ///
+    /// This is the union of index of the target basic block, the index assigned to the of the
+    /// target statement or terminator, and the indices of all preceding statements in the MIR.
+    ///
+    /// For example, the expected state when calling
+    /// `seek_before(Location { block: 2, statement_index: 2 })` would be `[102, 0, 1, 2, 3, 4]`.
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/tests.rs]: I spent a little while reading over the dummy_body()  cfg construction code and tried to understand why that control-flow graph would yield the results `[102, 0, 1, 2, 3, 4]`, but at some point I decided that if I have to work this hard to figure it out, then it needs to be explained more explicitly in the documentation here. In short, I don't know why I should expect the aforementioned output for that case. (e.g. Where does the 102 come from?)

File: src/librustc_mir/dataflow/generic/engine.rs [hunk 10]
```diff
@@ -167,124 +156,155 @@ where
         results
     }
 
-    /// Applies the cumulative effect of an entire block, excluding the call return effect if one
-    /// exists.
-    fn apply_whole_block_effect(
+    /// Helper function that propagates dataflow state into graph succesors until fixpoint is
+    /// reached.
+    fn _iterate_to_fixpoint(
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/generic/engine.rs]: eek, I don't think you should name helpers with a *leading* underscore. Leading underscore is usually used for fields/methods that might be dead code.

I'd personally use a word to differentiate the helper routiine (e.g. `iterate_to_fixpoint_core`) but even just a *trailing* underscore would be better here than a leading underscore.

**Comment by pnkfelix** [comment]
Okay. Overall this looks good. My main complaint is about the `impl GenKill for BitSet`. Are you using that impl? If not, then I'd say get rid of it (and potentially get rid of `trait GenKill` entirely, at least for now.

**Comment by pnkfelix** [comment]
Regarding the specific questions raised by @ecstatic-morse : 

> - How do we feel about the `{before_,}*_effect` naming convention? It takes a bit of getting used to, and doesn’t extend nicely to backwards dataflow.
>    - OTOH, having a canonical `statement_effect` which most implementers want to use and a secondary, prefixed effect which is only rarely used is good.

While reading the commit series, I definitely asked myself "can we get rid of the `before_{statement, terminator}_effect` methods entirely? Its not clear what purpose they currently serve, because they are always called (I think) immediately before their corresponding partner `{statement, terminator}_effect` method, right? (Or is there some context where only the latter is called without the former? I guess I would need to review the code more thoroughly.

In any case, I would be fine with leaving the convention as is for now, and making a follow-up task to revisit this part of the API. I'd be thrilled to clean it up, but I wouldn't block landing this PR on that task.

> - Should we be passing a `mir::Statement` to the `statement` effect methods? What about `mir::Body` to `bits_per_block`?
>    - I'd like to have a `HasBody<'mir, 'tcx>` trait that is a requirement for `AnalysisDomain` instead of passing in a separate body around everywhere, but this will increase the amount of boilerplate required.

I'm fine with passing in the `Statement`; it seems like most effects will need that info, so why make all the clients do the lookup, right? I don't care as much either way about whether or not we pass in the `Body` to `bits_per_block` (I don't know offhand, but I would assume a similar argument applies there that most consumers will need to extract info from the `Body` in order to determine the size of their domain's bitvector.)

> - Should the `Analysis` methods take `&self` or `&mut self`?
>     - Some may depend on the results of an earlier dataflow analysis and want to hold a cursor.
>     - However, transfer functions should be pure for a given MIR, so maybe forcing those analyses to use interior mutability is appropriate.

I'm torn. I would vote for `&self` here, personally. But I know first-hand the pain of trying to wedge in an analysis-dependent analysis after the framework was originally designed without that use case in mind, and your argument there for `&mut self` is at least giving me pause.

Is there some semi-standard way we could set things up so that callers of `Analysis` see its API as `&mut self`, while also having implementors of `Analysis` see its API as `&self`, so that we would force each side to be conservative in their implementation choices while delaying committing ourselves to one side or another right now?

File: src/librustc_mir/dataflow/generic/mod.rs (hunk 3 above)
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic/mod.rs]: In `ResultsCursor`, the `apply` methods are applied directly to the `BitSet` containing the dataflow state at the cursor position. The `GenKill` trait is meant to abstract over two different contexts where we need to apply the transfer function. When coalescing transfer functions for an entire block before running dataflow analysis, we are storing a function (`trans`) that can be applied later, which requires two bitsets. However, when seeking with a dataflow cursor, we don't care about the function itself; we only need to know its effect on the state vector, (`trans(state)`). There's no need to store extra bitsets.

There are other ways to avoid allocating here, but having a proper trait is the cleanest approach IMO. I agree that the documentation should be clearer about what is happening when in the `GenKill` implementation for `BitSet`.

File: src/librustc_mir/dataflow/generic/tests.rs (hunk 7 above)
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic/tests.rs]: I meant the first interpretation. This was unclear.

File: src/librustc_mir/dataflow/generic/tests.rs (hunk 6 above)
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic/tests.rs]: I'll start from first principles since I didn't do a good job of explaining the motivation for this.

The ultimate goal of this is to test the logic that updates the internal state of a `dataflow::ResultsCursor` (this is where indices should be "set") by seeking *to* every possible target (before or after a `Location`) *from* every possible target. For this test to be exhaustive, we must define a dataflow analysis such that every possible effect (before, after and call-return) is unique for every statement and terminator in a MIR body. This ensures that the expected state at every possible target is unique over the entire MIR body. There's no need to actually find the fixpoint for such an analysis to test `ResultsCursor`, so the actual CFG is inconsequential.

The actual implementation does not quite live up to this ideal. Instead it takes advantage of the fact that we never iterate to fixpoint by storing the index of each basic block in its entry set. The effect of each statement in our dummy analysis is only "unique" when considered in conjunction with the index that appears in the entry set.

I could reimplement this test it to be more like the first paragraph. Since I need to map before/after/call-return effects for each statement and terminator to a unique index, I'm afraid that it will always be a bit opaque. I could also keep it roughly as-is with an edited version of this comment as documentation, or remove it entirely.

**Comment by ecstatic-morse** [comment]
> While reading the commit series, I definitely asked myself "can we get rid of the before_{statement, terminator}_effect methods entirely? Its not clear what purpose they currently serve, because they are always called (I think) immediately before their corresponding partner {statement, terminator}_effect method, right? (Or is there some context where only the latter is called without the former? I guess I would need to review the code more thoroughly.)

The only implementer of the "before" variants is in [`src/librustc_mir/dataflow/impls/borrows.rs`](https://github.com/rust-lang/rust/blob/25434f898b499876203a3b95c1b38bad5ed2cc5d/src/librustc_mir/dataflow/impls/borrows.rs#L249). They could almost be factored out by incorporating the "before" effect of statement `i` into the "after" effect of statement `i-1`. However, this won't work if the first statement in a block has a meaningful "before" effect. This may not be the case for the `Borrows` analysis, since what loans would go out of scope before the first statement?

File: src/librustc_mir/dataflow/generic/engine.rs (hunk 10 above)
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic/engine.rs]: This is my personal coding convention that probably originates from Python. I will switch this (as well as `_seek`) to a trailing underscore unless there is an established convention.

**Comment by ecstatic-morse** [comment]
Ideally, this PR would be performance neutral, but there's a slight regression on [two benchmarks](https://perf.rust-lang.org/compare.html?start=76a252ea9e7be93a61ffdf33b3533e24a9cf459d&end=d05d26499af24c8153dd8f16bb860abe185d18b2). I currently have a micro-optimization in this PR that partially addresses this. I would like to split that optimization out into a later PR that also ports existing analyses to the new framework (importantly `MaybeInitializedPlaces`, `MaybeUninitializedPlaces`, `Borrows`) to the new framework so I can get a fuller picture of its performance. This PR only includes a port of the first, which means the perf diff is more susceptible to noise.

**Comment by ecstatic-morse** [comment]
> 

Is there some semi-standard way we could set things up so that callers of Analysis see its API as &mut self, while also having implementors of Analysis see its API as &self, so that we would force each side to be conservative in their implementation choices while delaying committing ourselves to one side or another right now?

I suppose we could have an `AnalysisMut` facade that is used by the `engine` and `cursor`. I think it's not worth the extra complexity. I'd rather just keep the status quo and fix any borrow check errors in `engine` and `cursor` by hand if we decide we want to switch to `&mut analysis`.

**Comment by ecstatic-morse** [comment]
@pnkfelix I reorganized the unit test and added docs to `GenKill` and `GenKillSet` to explain how they should be used. I've moved the migration of `MaybeInitializedPlaces` to the new framework into a separate branch so that I can work on optimizing them without pushing a bunch of commits to this PR.

**Comment by ecstatic-morse** [comment]
I migrated the analyses used during borrow checking/drop elaboration to the new framework and perf [seems good](https://github.com/rust-lang/rust/pull/68241#issuecomment-574570513). I'm gonna try a few small optimizations I mentioned during the design meeting over on that PR as well.

While doing the conversion, it became apparent that code would much be nicer if we extended the "builder" style all the way back to `Analysis` by adding an `into_engine` method to it. For example, code like this:

```rust
let flow_inits = MaybeInitializedPlaces::new(tcx, &body, &mdpe);
let mut flow_inits = dataflow::generic::Engine::new_gen_kill(tcx, &body, def_id, flow_inits)
        .iterate_to_fixpoint()
        .into_cursor(&body);
```

could instead be written like this:

```rust
let mut flow_inits = MaybeInitializedPlaces::new(tcx, &body, &mdpe)
        .into_engine(tcx, &body, def_id)
        .iterate_to_fixpoint()
        .into_cursor(&body);
```

Another idea was to add `HasBody` as a supertrait of `AnalysisDomain`. This would obviate the need to pass `body` everywhere. With the current API, it's possible (albeit unlikely) for a user to pass a different `mir::Body` into, for example, `into_engine` than was used to create the analysis, which would cause subtle bugs. I believe this would only work if the `Analysis` methods used `&self`, however.

**Comment by ecstatic-morse** [comment]
@pnkfelix I addressed the motivation for the `GenKill` trait and added some documentation/renamed some methods in the last two commits.

**Comment by pnkfelix** [comment]
Thanks @ecstatic-morse !

File: src/librustc_mir/dataflow/generic/mod.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc_mir/dataflow/generic/mod.rs]: Okay. I'm sorry that I did not read this response more carefully back when you first posted it, as it would have saved me the effort of rediscovering the motivation for this abstraction myself. (But then again, it did force you to embed the motivation into the doc-comment for the trait, so maybe my oversight was a win in the long run...)
